hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/HoodieTimelineArchiver.java

```diff
@@ -452,107 +431,137 @@ private Stream<HoodieInstant> getCommitInstantsToArchive() throws IOException {
-                  .orElse(true)
-          ).filter(s ->
-              oldestInstantToRetainForClustering.map(instantToRetain ->
-                      HoodieTimeline.compareTimestamps(s.getTimestamp(), LESSER_THAN, instantToRetain.getTimestamp()))
-                  .orElse(true)
-          );
-      return instantToArchiveStream.limit(commitTimeline.countInstants() - minInstantsToKeep);
+          }).filter(s -> oldestInstantToRetain
+              .map(instant -> compareTimestamps(s.getTimestamp(), LESSER_THAN, instant.getTimestamp()))
+              .orElse(true));
+      return instantToArchiveStream.limit(completedCommitsTimeline.countInstants() - minInstantsToKeep)
+          .collect(Collectors.toList());
     } else {
-      return Stream.empty();
+      return Collections.emptyList();
     }
   }
 
+  private List<HoodieInstant> getCleanAndRollbackInstantsToArchive(HoodieInstant newestCommitInstantToArchive) {
+    HoodieTimeline cleanAndRollbackTimeline = table.getActiveTimeline()
+        .getTimelineOfActions(CollectionUtils.createSet(HoodieTimeline.CLEAN_ACTION, HoodieTimeline.ROLLBACK_ACTION))
+        .filterCompletedInstants();
+
+    // Since the commit instants to archive is continuous, we can use the newest commit instant to archive as the
+    // right boundary to collect the clean or rollback instants to archive.
+    //
+    //                                                  newestCommitInstantToArchive
+    //                                                               v
+    //  | commit1 clean1 commit2 commit3 clean2 commit4 rollback1 commit5 | commit6 clean3 commit7 ...
+    //  | <------------------  instants to archive  --------------------> |
+    //
+    //  CommitInstantsToArchive: commit1, commit2, commit3, commit4, commit5
+    //  CleanAndRollbackInstantsToArchive: clean1, clean2, rollback1
+
+    return cleanAndRollbackTimeline.getInstantsAsStream()
+        .filter(s -> compareTimestamps(s.getTimestamp(), LESSER_THAN, newestCommitInstantToArchive.getTimestamp()))
+        .collect(Collectors.toList());
+  }
+
   private Stream<HoodieInstant> getInstantsToArchive() throws IOException {
-    Stream<HoodieInstant> instants = Stream.concat(getCleanInstantsToArchive(), getCommitInstantsToArchive());
     if (config.isMetaserverEnabled()) {
       return Stream.empty();
     }
 
-    // For archiving and cleaning instants, we need to include intermediate state files if they exist
+    // First get commit instants to archive.
+    List<HoodieInstant> instantsToArchive = getCommitInstantsToArchive();
+    if (!instantsToArchive.isEmpty()) {
+      HoodieInstant newestCommitInstantToArchive = instantsToArchive.get(instantsToArchive.size() - 1);
+      // Then get clean and rollback instants to archive.
+      List<HoodieInstant> cleanAndRollbackInstantsToArchive =
+          getCleanAndRollbackInstantsToArchive(newestCommitInstantToArchive);
+      instantsToArchive.addAll(cleanAndRollbackInstantsToArchive);
```

Comment:
Do we need to sort all these instants.

Reply:
It was also unsorted before this PR. But I can add a sort here, obviously the instants written to the archive should be ordered